### DISCUSSION TITLE
Reproducing ./echo functionality with ./echo-non-kens

### DISCUSSION THREAD

**Discussion by dormail** [ANLAB-KAIST/KENSv3#24] [Project Q&A 1. Application Layer (Socket)]
Is there an option to reproduce ./echo functionality with the ./echo-non-kens and some options? My client and server are giving correct responds but the tests are failing, e.g. one output could be
```
./echo-non-kens s 127.0.0.1 9000 echo-test
Server received
hello
Submitted answers:
  127.0.0.1 -> echo-test

./echo-non-kens c 127.0.0.1 9000 hello
Submitted answers:
  127.0.0.1 -> hello
```
and KENS is telling me
```
Expected equality of these values:
  server_ans
    Which is: ""
  expected_server_ans
    Which is: "echo-test"
```
Maybe I also did not understand the submitAnswer method, since it's purpose is not really clear to me (I used read/write to send/receive the data).

**Comment by Pusnow** [comment]
In the server application, you should call `submitAnswer` with **client's IP address** and **request's content** not response's.
In the client application, you should call `submitAnswer` with **server's  IP address** and **response's content** not request's.

It seems you did the opposite.

**Reply by dormail** [reply]
Ok so when I would have a server consisting only out of
```
int servermain(....) {
  submitAnswer("10.0.0.1", "echo-test");
  return 0;
}
```
it should at least pass the first test, right? The whole submitAnswer and ./echo feedback looks like some dark magic to me.

**Reply by Pusnow** [reply]
No, since the client's IP address is 10.0.0.2. It should be `submitAnswer("10.0.0.2", "echo-test");`.

**Reply by dormail** [reply]
Ok I have done the same with the IP you mentioned
```
int EchoAssignment::serverMain(const char *bind_ip, int port,
                               const char *server_hello) {
  submitAnswer("10.0.0.2", "echo-test");
  return 0;
}
```
and ./echo still does not show anything. What I have noticed is that `./echo-non-kens s` only works when using the 127.0.0.x, not with the 10.0.0.x adresses, because the Socket could not be bound. From the errno it's clear that the 10.0.0.x can not be used because it is not  local. I am on Debian 11, is it possible that other distros do not have this problem?

**Reply by songminwoo1** [reply]
I think I'm having a similar problem. My program works perfectly OK with ./echo-non-kens, but if I use ./echo, the program just don't receive any data at all.

**Reply by Pusnow** [reply]
@dormail It's totally expected. It is not possible to "bind" unavailable IPs.

To test with real network, you should bind 127.0.0.1 or your machine's IP address. 0.0.0.0 (bind all available IPs) is also file.

**Reply by Pusnow** [reply]
@songminwoo1 Could you explain more?

**Reply by Jongjunp** [reply]
I also suffer from the same problem.
When I run my code with './echo-non-kens', it shows me the following result.
For client, 
`./echo-non-kens c 127.0.0.1 9000 hello`
`Submitted answers:
  127.0.0.1 -> nice_to_meet_you`
For server, 
`./echo-non-kens s 127.0.0.1 9000 nice_to_meet_you`
`Submitted answers:
  127.0.0.1 -> hello`

However, when I run my code with './echo', it returns the following result.
For client, 
`: Failure
Expected equality of these values:
  client_ans
    Which is: ""
  expected_client_ans
    Which is: "echo-test"`
For server,
`: Failure
Expected equality of these values:
  server_ans
    Which is: ""
  expected_server_ans
    Which is: "echo-test"`

I can't guess what is wrong with my code from the message.

**Reply by idev2580** [reply]
I also suffering from same problem with @Jongjunp.
KENS doesn't get any answers from my program. (Without KENS, it works perfectly.)

I tested 
```
int EchoAssignment::serverMain(const char *bind_ip, int port,
                               const char *server_hello) {
  submitAnswer("KENS", "NOT WORK AT ALL");
  return 0;
}
```
but `./echo` returned following result.
`: Failure Expected equality of these values:
  server_ans
    Which is: ""
  expected_server_ans
    Which is: "echo-test"`
(even though actual answer my code submitted was "NOT WORK AT ALL")

**Reply by Pusnow** [reply]
@Jongjunp @Xxix2580 see below answer.

**Comment by Pusnow** [comment]
For all who have message like:

> Failure Expected equality of these values: client_ans Which is: "" expected_client_ans Which is: "echo-test"

If the your client/server's answer is empty (""), you may submit answer with wrong IP address.

This is `submitAnswer` method.

https://github.com/ANLAB-KAIST/KENSv3/blob/3098c763bdc84e3c2f8852040113cfd76639e5f0/app/echo/EchoAssignment.hpp#L51-L55

You must submit answer with peer's IP address.
In the server, you should submit answer with client's IP address.
In the client, you should submit answer with server's IP address.

To know peer's IP address, you should use [`getpeername(2)`](https://man7.org/linux/man-pages/man2/getpeername.2.html).

**Reply by dormail** [reply]
That is what I am doing the whole time already. My Server's response boils down to
1. write(response)
2. getting client's IP address with `getpeername()` and transforming it into a char array
3. submitAnswer(clientsIP, request)

I also confirmed that it is the right IP address with outputs in the client and server program, they work perfectly in the ./echo-non-kens but nothing happens in ./echo

**Reply by Jongjunp** [reply]
I think your socket is declared as 
`socket(AF_INET, SOCK_STREAM, 0)`
In KENS, you should declare a socket as follows.
`socket(AF_INET, SOCK_STREAM, IPPROTO_TCP)`
See [https://github.com/ANLAB-KAIST/KENSv3/discussions/29](url)

**Reply by dormail** [reply]
@Jongjunp Yeah I have seen the discussion and changed the socket input parameter. Now the ./echo does not halt anymore, but the output still does not show any submitted answers.

**Reply by Pusnow** [reply]
@dormail okay, could you send your codes to us? cs341.ta.pj@kaist.ac.kr

**Reply by Pusnow** [reply]
@dormail I've checked your source codes and found you didn't initialize `addrlen_cl` (the third parameter of accept).

According to [accept(2)](https://man7.org/linux/man-pages/man2/accept.2.html), you should initialize it to size of the `address_cl` (the second parameter).
> The addrlen argument is a value-result argument: the caller must initialize it to contain the size (in bytes) of the structure pointed to by addr; on return it will contain the actual size of the peer address.

**Reply by dormail** [reply]
Yes, now it works, thank you very much for your efforts!
